internal/pkg/db/redis/data.go

```diff
@@ -458,6 +588,114 @@ func (c *Client) DeleteReadingById(id string) error {
+func (c *Client) DeleteReadingsByDevice(device string) error {
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.ReadingsCollection+":device:"+device, 0, -1))
+	if err != nil {
+		return err
+	}
+
+	err = conn.Send("MULTI")
+	if err != nil {
+		return err
+	}
+
+	for _, id := range ids {
+		err = conn.Send("RENAME", id, db.DeletedReadingsCollection+":"+id)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = conn.Send("EXEC")
+	deleteReadingsChannel <- device
+
+	return nil
+}
+
+// AsyncDeleteReadings Handles the deletion of device readings asynchronously. This function is expected to be running
+// in a go-routine and works with the "DeleteReadingsByDevice" function for better performance.
+func (c *Client) AsyncDeleteReadings(loggingClient logger.LoggingClient) {
+	loggingClient.Debug("Starting background event deletion process")
+	for {
+		select {
+		case device, ok := <-deleteReadingsChannel:
+			if ok {
+				loggingClient.Debug("Deleting reading data for device: " + device)
+				startTime := time.Now()
+				c.deleteRenamedReadings(device, loggingClient)
+				loggingClient.Debug(fmt.Sprintf("Deleted readings for device: '%s', elapsed time: %s", device, time.Since(startTime)))
+			}
+		}
+	}
+}
+
+// deleteRenamedReadings deletes all readings associated with the specified device which have been marked for deletion.
+// See the "DeleteReadingsByDevice" function for details on the how readings are marked for deletion(renamed)
+func (c *Client) deleteRenamedReadings(device string, loggingClient logger.LoggingClient) {
+	conn := c.Pool.Get()
+	defer conn.Close()
+
+	ids, err := redis.Strings(conn.Do("ZRANGE", db.ReadingsCollection+":device:"+device, 0, -1))
+	if err != nil {
+		loggingClient.Error("Unable to delete reading:" + err.Error())
+		return
+	}
+
+	_, err = conn.Do("MULTI")
+	if err != nil {
+		loggingClient.Error("Unable to start transaction for deletion:" + err.Error())
+	}
+
+	for _, id := range ids {
+		_, err = conn.Do("GET", db.DeletedReadingsCollection+":"+id)
+		if err != nil {
+			loggingClient.Error("Unable to obtain readings marked for deletion:" + err.Error())
+		}
+	}
+	readings, err := redis.Strings(conn.Do("EXEC"))
+
+	queriesInQueue := 0
+	var r contract.Reading
+	_, err = conn.Do("MULTI")
+	if err != nil {
+		loggingClient.Error("Unable to start batch processing for reading deletion:" + err.Error())
+	}
+
+	for _, reading := range readings {
+		err = unmarshalObject([]byte(reading), &r)
+		if err != nil {
+			loggingClient.Error("Unable to marshal reading: " + err.Error())
+		}
+		_ = conn.Send("UNLINK", db.DeletedReadingsCollection+":"+r.Id)
+		_ = conn.Send("ZREM", db.ReadingsCollection, r.Id)
+		_ = conn.Send("ZREM", db.ReadingsCollection+":created", r.Id)
+		_ = conn.Send("ZREM", db.ReadingsCollection+":device:"+r.Device, r.Id)
+		_ = conn.Send("ZREM", db.ReadingsCollection+":name:"+r.Name, r.Id)
+		queriesInQueue++
+
+		if queriesInQueue >= AsyncDeleteBatchSize {
```

Comment:
Suggest `c. AsyncDeleteBatchSize` driven from configuration as mentioned in comment above -- tune-able as performance needs / resources warrant.